src/Workspaces/Core/Portable/Shared/Utilities/BloomFilter.cs

```diff
@@ -395,4 +394,74 @@ private static bool IsEquivalent(BitArray array1, BitArray array2)
 
         return true;
     }
+
+    /// <summary>
+    /// Provides mechanism to efficiently obtain bloom filter hash for a value. Backed by a single element cache.
+    /// </summary>
+    internal sealed class BloomFilterHash
+    {
+        private static BloomFilterHash? s_cachedHash;
+
+        private readonly string _value;
+        private readonly bool _isCaseSensitive;
+        private readonly ImmutableArray<int> _hashes;
+
+        private BloomFilterHash(string value, bool isCaseSensitive, int hashFunctionCount)
+        {
+            _value = value;
+            _isCaseSensitive = isCaseSensitive;
+
+            var hashBuilder = new FixedSizeArrayBuilder<int>(hashFunctionCount);
+
+            for (var i = 0; i < hashFunctionCount; i++)
+                hashBuilder.Add(BloomFilter.ComputeHash(value, i, _isCaseSensitive));
+
+            _hashes = hashBuilder.MoveToImmutable();
+        }
+
+        /// <summary>
+        /// Although calculating this hash isn't terribly expensive, it does involve multiple
+        /// (usually around 13) hashings of the string (the actual count is <see cref="BloomFilter._hashFunctionCount"/>).
+        /// The typical usage pattern of bloom filters is that some operation (eg: find references)
+        /// requires asking a multitude of bloom filters whether a particular value is likely contained.
+        /// The vast majority of those bloom filters will end up hashing that string to the same values, so
+        /// we put those values into a simple cache and see if it can be used before calculating.
+        /// Local testing has put the hit rate of this at around 99%.
+        /// </summary>
+        public static ImmutableArray<int> GetOrCreateHashArray(string value, bool isCaseSensitive, int hashFunctionCount)
```

Comment:
this should mention you can get a larger array, but should only use up to the first hashFunctionCount entries in it.